### PR TITLE
#3719 Add reason column for vaccines

### LINKED ISSUE #3719: Allow users to choose an open vial wastage reason for vaccines in a stock take
## Is your feature request related to a problem? Please describe.

We need to allow for a user to select an open vial wastage reason in a stocktake.

## Describe the solution you'd like

For a vaccine item in a stoctake, allow a user to select `negativeInventoryAdjustment` OR `openVialWastage` reasons for a vaccine item with a negative difference.

## Implementation

Unfortunately, there's no room. 

I think maybe we need to remove the breach column..?

## Describe alternatives you've considered

N/A

## Additional context

N/A

### PR DESCRIPTION
**Pull request by joshxg**
Fixes #3719 

## Change summary

- Adds a reason column to vaccines, and allows for applying an `openVialWastage`
- Quite a hacky solution. Not very happy about it, but 

## Testing

- [ ] Can apply a reason to a vaccine in a stocktake

### Related areas to think about

image

### CODE REVIEW

File: src/widgets/modals/StocktakeBatchModal.js [hunk 1]
```diff
@@ -89,14 +96,26 @@ export const StocktakeBatchModalComponent = ({
...
-    difference > 0
-      ? UIDatabase.objects('PositiveAdjustmentReason')
-      : UIDatabase.objects('NegativeAdjustmentReason');
+  let reasonsSelection = null;
+  if (usingReasons && !isVaccine) {
+    reasonsSelection =
+      difference > 0
+        ? UIDatabase.objects('PositiveAdjustmentReason')
+        : UIDatabase.objects('NegativeAdjustmentReason');
+  } else if (usingOpenVialWastageReasons && isVaccine) {
```
Comment (mark-prins) [inline src/widgets/modals/StocktakeBatchModal.js:109]: It is possible that `usingReasons` is false and it is a vaccine and there aren't any positive adjustment reasons, only the wastage reasons.

Though it really doesn't matter as the `REASON` column only shows if there are positive and negative reasons available. Really just wondering if the query could be simplified at all.

File: src/widgets/modals/StocktakeBatchModal.js [hunk 2]
```diff
@@ -89,14 +96,26 @@ export const StocktakeBatchModalComponent = ({
...
+    reasonsSelection =
+      difference > 0
+        ? UIDatabase.objects('PositiveAdjustmentReason')
+        : UIDatabase.objects('Options').filtered(
+            '(type == $0 || type == $1) && isActive == true',
+            'negativeInventoryAdjustment',
+            'openVialWastage'
+          );
+  }
```
Comment (mark-prins) [inline src/widgets/modals/StocktakeBatchModal.js:118]:
```suggestion
  const reasonsSelection = (() => {
    if (difference > 0) {
      return UIDatabase.objects('PositiveAdjustmentReason');
    }

    if (usingReasons && !isVaccine) {
      return UIDatabase.objects('NegativeAdjustmentReason');
    }

    if (usingOpenVialWastageReasons && isVaccine) {
      return UIDatabase.objects('Options').filtered(
        '(type == $0 || type == $1) && isActive == true',
        'negativeInventoryAdjustment',
        'openVialWastage'
      );
    }

    return [];
  })();
```

**Review by mark-prins — APPROVED** [review]
that looks reasonable 😉 

all good - as you say, could perhaps be neater in the UI but works fine.

File: src/widgets/modals/StocktakeBatchModal.js (hunk 2 above)
Comment (joshxg) [reply src/widgets/modals/StocktakeBatchModal.js]: oooo nice fancy pants

File: src/widgets/modals/StocktakeBatchModal.js (hunk 1 above)
Comment (joshxg) [reply src/widgets/modals/StocktakeBatchModal.js]: Yeah, it's not great and I don't know what to do about it.. :(

**Pull request merged**